src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,1006 @@
+        int extension_size = 0;
+        int i;
+        char *buf;
+        int size_written;
+
+        for (i = 0 ; i < rd_list_cnt(&state->extensions) ; i++) {
+                rd_strtup_t *extension = rd_list_elem(&state->extensions, i);
+                // kvpair         = key "=" value kvsep
+                extension_size += strlen(extension->name) + 1 // "="
+                        + strlen(extension->value) + kvsep_size;
+        }
+
+        // client-resp    = (gs2-header kvsep *kvpair kvsep) / kvsep
+        out->size = strlen(gs2_header) + kvsep_size
+                + strlen("auth=Bearer ") + strlen(state->token_value)
+                + kvsep_size + extension_size + kvsep_size;
+        out->ptr = rd_malloc(out->size+1);
+
+        buf = out->ptr;
+        size_written = 0;
+        size_written += rd_snprintf(buf, out->size+1 - size_written,
+                "%s%sauth=Bearer %s%s", gs2_header, kvsep, state->token_value,
+                kvsep);
+        buf = out->ptr + size_written;
+
+        for (i = 0 ; i < rd_list_cnt(&state->extensions) ; i++) {
+                rd_strtup_t *extension = rd_list_elem(&state->extensions, i);
+                size_written += rd_snprintf(buf, out->size+1 - size_written,
+                        "%s=%s%s", extension->name, extension->value, kvsep);
+                buf = out->ptr + size_written;
+        }
+
+        rd_snprintf(buf, out->size+1 - size_written, "%s", kvsep);
+
+        rd_rkb_dbg(rktrans->rktrans_rkb, SECURITY, "SASLOAUTHBEARER",
+                   "Built client first message");
+}
+
+
+
+/**
+ * @brief SASL OAUTHBEARER client state machine
+ * @returns -1 on failure (errstr set), else 0.
+ */
+static int rd_kafka_sasl_oauthbearer_fsm (rd_kafka_transport_t *rktrans,
+                                    const rd_chariov_t *in,
+                                    char *errstr, size_t errstr_size) {
+        static const char *state_names[] = {
+                "client-first-message",
+                "server-first-message",
+                "server-failure-message",
+        };
+        struct rd_kafka_sasl_oauthbearer_state *state =
+                rktrans->rktrans_sasl.state;
+        rd_chariov_t out = RD_ZERO_INIT;
+        int r = -1;
+
+        rd_rkb_dbg(rktrans->rktrans_rkb, SECURITY, "SASLOAUTHBEARER",
+                   "SASL OAUTHBEARER client in state %s",
+                   state_names[state->state]);
+
+        switch (state->state)
+        {
+        case RD_KAFKA_SASL_OAUTHB_STATE_SEND_CLIENT_FIRST_MESSAGE:
+                rd_dassert(!in); /* Not expecting any server-input */
+
+                rd_kafka_sasl_oauthbearer_build_client_first_message(rktrans,
+                        &out);
+                state->state =
+                        RD_KAFKA_SASL_OAUTHB_STATE_RECV_SERVER_FIRST_MSG;
+                break;
+
+
+        case RD_KAFKA_SASL_OAUTHB_STATE_RECV_SERVER_FIRST_MSG:
+                if (!in->size || !*in->ptr) {
```

Comment:
no need to check in->ptr

Reply:
We actually receive an empty string from the broker in the success case, and this presents as `in->size` = 4 and `*in->ptr` = '\0\'.  We receive a JSON  object with an error message upon failure, which presents as a non-zero size and a non-empty string, so I think we should keep it the way it is.

Reply:
I stepped through in the debugger, and it definitely presents as `in->size` = 4 and `*in->ptr` = '\0'.  The broker is sending an empty string back.  https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java#L182.